Improve decimal aggregations

martint opened the pull request:
Remove unnecessary calls to get state, which can be expensive due to
lookups in the underlying BigArrays.

https://jmh.morethan.io/?sources=https://gist.githubusercontent.com/martint/f2685c5b8fc19f8feaa2a3ea03c43bda/raw/82b40a4efcbae857f901e9f51105902790cbf036/before.json,https://gist.githubusercontent.com/martint/f2685c5b8fc19f8feaa2a3ea03c43bda/raw/82b40a4efcbae857f901e9f51105902790cbf036/after.json

pettyjamesm commented on core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:
```diff
@@ -161,27 +161,34 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
         }
+        long overflow = UnscaledDecimal128Arithmetic.addWithOverflow(sum, unscaledValue, sum);
+        if (overflow != 0) {
+            state.setOverflow(state.getOverflow() + overflow);
```
Seems like `state.addOverflow(overflow)` would suffice here.

martint replied:
oops... yeah, missed this one

pettyjamesm commented on core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:
```diff
@@ -161,27 +161,34 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
-        if (state.getLongDecimal() == null) {
-            state.setLongDecimal(UnscaledDecimal128Arithmetic.unscaledDecimal());
+        Slice sum = state.getLongDecimal();
+        if (sum == null) {
```
See the below comment about maybe having and using something like:
 ```
 Slice sum = state.getOrInitializeLongDecimal(unscaledValue);
 if (sum != unscaledValue) {
   state.addOverflow(UnscaledDecimal128Arithmetic.addWithOverflow(sum, unscaledValue, sum));
 }
 ```

pettyjamesm replied:
Disregard, this strategy is ok for short decimal where we know it was just heap allocated but extremely dangerous for long decimals where you might have a Slice pointing into the middle of the input block.

pettyjamesm commented on core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:
```diff
@@ -161,27 +161,34 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
+        }
+
+        if (overflow != 0) {
+            state.addOverflow(overflow);
```
Disregard, this (appears) to be material in my local benchmarking. Doing the check inside of the add method happens after bimorphic method dispatch and is more expensive than localizing the check AFAICT.

martint commented on core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:
```diff
@@ -161,27 +161,34 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
-        if (state.getLongDecimal() == null) {
+        long overflow = otherState.getOverflow();
+
+        Slice sum = state.getLongDecimal();
```
I'm thinking of another follow up change with a dedicated implementation of the aggregates for long vs short decimals, where we avoid using Slice entirely for short decimals.

martint committed:
Improve aggregation performance for decimal

Remove unnecessary calls to get state, which can be expensive due to
lookups in the underlying BigArrays.

before

Benchmark                              (function)  (groupCount)  (type)  Mode  Cnt   Score   Error  Units
BenchmarkDecimalAggregation.benchmark         avg            10   SHORT  avgt   30  19.597 ± 1.962  ns/op
BenchmarkDecimalAggregation.benchmark         avg            10    LONG  avgt   30  21.407 ± 1.198  ns/op
BenchmarkDecimalAggregation.benchmark         avg          1000   SHORT  avgt   30  23.303 ± 1.261  ns/op
BenchmarkDecimalAggregation.benchmark         avg          1000    LONG  avgt   30  23.077 ± 1.469  ns/op
BenchmarkDecimalAggregation.benchmark         sum            10   SHORT  avgt   30  19.253 ± 0.902  ns/op
BenchmarkDecimalAggregation.benchmark         sum            10    LONG  avgt   30  19.402 ± 0.990  ns/op
BenchmarkDecimalAggregation.benchmark         sum          1000   SHORT  avgt   30  21.180 ± 0.942  ns/op
BenchmarkDecimalAggregation.benchmark         sum          1000    LONG  avgt   30  21.834 ± 2.131  ns/op

after

Benchmark                              (function)  (groupCount)  (type)  Mode  Cnt   Score   Error  Units
BenchmarkDecimalAggregation.benchmark         avg            10   SHORT  avgt   30  14.502 ± 0.387  ns/op
BenchmarkDecimalAggregation.benchmark         avg            10    LONG  avgt   30  14.967 ± 0.525  ns/op
BenchmarkDecimalAggregation.benchmark         avg          1000   SHORT  avgt   30  16.087 ± 0.557  ns/op
BenchmarkDecimalAggregation.benchmark         avg          1000    LONG  avgt   30  16.562 ± 0.611  ns/op
BenchmarkDecimalAggregation.benchmark         sum            10   SHORT  avgt   30  16.680 ± 0.730  ns/op
BenchmarkDecimalAggregation.benchmark         sum            10    LONG  avgt   30  16.498 ± 1.010  ns/op
BenchmarkDecimalAggregation.benchmark         sum          1000   SHORT  avgt   30  17.721 ± 0.513  ns/op
BenchmarkDecimalAggregation.benchmark         sum          1000    LONG  avgt   30  18.043 ± 1.145  ns/op

skrzypo987 commented on core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:
```diff
@@ -0,0 +1,157 @@
+@Warmup(iterations = 10)
+@Measurement(iterations = 10)
+@BenchmarkMode(Mode.AverageTime)
+public class BenchmarkDecimalAggregation
```
You can move that to a separate commit and post before/after results in this commit message

pettyjamesm commented on core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:
```diff
@@ -211,19 +214,11 @@ public static BigDecimal average(LongDecimalWithOverflowAndLongState state, Deci
             BigInteger overflowMultiplier = TWO.shiftLeft(UNSCALED_DECIMAL_128_SLICE_LENGTH * 8 - 2);
-            BigInteger overflow = overflowMultiplier.multiply(BigInteger.valueOf(state.getOverflow()));
-            sum = sum.add(new BigDecimal(overflow));
+            sum = sum.add(new BigDecimal(overflowMultiplier.multiply(BigInteger.valueOf(overflow))));
```
Overflow multiplier is a constant

martint committed:
Extract constant

martint committed:
Add benchmark for decimal aggregations

sopel39 commented:
[Uploading Benchmarks comparison-aggrdec.pdf…]()

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:
```diff
@@ -161,28 +161,31 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
     {
-        state.setLong(state.getLong() + otherState.getLong());
-        state.setOverflow(state.getOverflow() + otherState.getOverflow());
+        state.addLong(otherState.getLong());
```
it would be nice to have some comments what `addLong` means (e.g updating row counter)

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:
```diff
@@ -161,28 +161,31 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
+        Slice sum = state.getLongDecimal();
+        if (sum == null) {
+            sum = UnscaledDecimal128Arithmetic.unscaledDecimal();
+            state.setLongDecimal(sum);
```
you could skip `addWithOverflow` completely here as in `combine` method

martint replied:
Unfortunately, no. The way this aggregation manages state is a little messy. It relies on side effects to update the state: 
* it sets the state to an empty slice if it's null
* if relies on addWithOverflow to update it in place

I think there are opportunities to clean this up and improve it later, but it will require a rewrite and specialization for short vs long decimals

sopel39 replied:
You could use `state.setLongDecimal(UnscaledDecimal128Arithmetic#unscaledDecimal(unscaledValue))` (it does copy slice)
Are we afraid that original `unscaledValue` will be modified?

martint replied:
I didn't change the original logic (which I think is confusing and hard to reason about) -- I just restructured the code to avoid repeated calls to get state.

I'll leave that for a future revamp of this class.

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:
```diff
@@ -161,28 +161,31 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
         }
+        long overflow = UnscaledDecimal128Arithmetic.addWithOverflow(sum, unscaledValue, sum);
+        state.addOverflow(overflow);
+        state.addLong(1);
```
you could move this before `if`

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalSumAggregation.java:
```diff
@@ -140,38 +140,35 @@ private static InternalAggregationFunction generateAggregation(Type inputType, T
+        Slice sum = state.getLongDecimal();
+        if (sum == null) {
+            sum = UnscaledDecimal128Arithmetic.unscaledDecimal();
+            state.setLongDecimal(sum);
```
you could skip `addWithOverflow` in this branch as in `combine`

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowStateFactory.java:
```diff
@@ -133,6 +141,12 @@ public void setOverflow(long overflow)
 
+        @Override
+        public void addOverflow(long overflow)
+        {
```
why no `if (overflow != 0) {` check?

martint replied:
Because the addition is cheap. In the grouped case, we guard it because the cause to read/set the value in the underlying array is more expensive

sopel39 reviewed (commented):
small comments

martint commented:
Updated

sopel39 commented on core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java:
```diff
@@ -161,28 +161,33 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
-        if (state.getLongDecimal() == null) {
+        Slice sum = state.getLongDecimal();
+        if (sum == null) {
             state.setLongDecimal(otherState.getLongDecimal());
```
now updates in place will also affect `otherState.getLongDecimal()`. Should we make a copy here or is `otherState` discarded after `combine`?

martint replied:
Same as above... I didn't change the original logic. I just restructured the code to avoid repeated calls to get state.

I'll leave that for a future revamp of this class.

pettyjamesm replied:
It’s risky and this is one of those cases where using slices that might be mutated it might be pointers into the middle of a very large allocation and therefore retain it (eg: VariableWidthBlock, the serialized block type of this state) are scary.

My understanding of this and the rest of the code is that this is safe because otherState is dead after combined with this state, and since the addInput call creates a new slice and doesn’t borrow from the input block when the current state is null, the second concern is also not a problem here. That said, this is indeed a fragile interaction. Not any more fragile than it was before, but hopefully something that can be improved with the future work.

sopel39 reviewed (approved):
lgtm % https://github.com/trinodb/trino/pull/6591#discussion_r558686781

pettyjamesm commented on core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalSumAggregation.java:
```diff
@@ -140,38 +140,41 @@ private static InternalAggregationFunction generateAggregation(Type inputType, T
         else {
-            accumulateValueInState(otherState.getLongDecimal(), state);
+            long overflow = UnscaledDecimal128Arithmetic.addWithOverflow(sum, otherState.getLongDecimal(), sum);
+            state.addOverflow(overflow);
```
Shouldn't this be: `state.addOverflow(overflow + otherState.getOverflow())`

martint commented:
> lgtm % #6591 (comment)

For some reason, that shape of code kills all the perf benefits. Probably some escape analysis issue that's getting in the way.
